vendor/github.com/containerd/nydus-snapshotter/pkg/converter/types.go

```diff
@@ -125,6 +125,8 @@ type MergeOption struct {
 	Timeout *time.Duration
 	// Encrypt encrypts the bootstrap layer if it's specified.
 	Encrypt Encrypter
+	// AppendFiles specifies the files that need to be appended to the bootstrap layer.
+	AppendFiles []File
```

Comment:
I was curious what this was for; see it was added in this PR;

- https://github.com/containerd/nydus-snapshotter/pull/591

Not sure how it's controlled though; the mention of "user can add some files" scared me a bit; do you have more context on this @imeoer  (from a security perspective?)

Reply:
Having a quick look, and it looks like this may be a continuation of https://github.com/containerd/nydus-snapshotter/commit/a3fccd09a979d46c6683257c7cc9f7ccdf75bb82, which (IIUC) allows for a custom image to be injected in the rootfs, for accelerated performance (https://github.com/containerd/nydus-snapshotter/issues/244)

> It allows nydus lazily load OCI targz layer with the nydus zran ref image.

So the previous code would allow for a "nydus zran" bootstrap image / tar to be used, from which the `EntryBootstrap` bootstrap-image would be loaded an injected into the init layer;
https://github.com/moby/buildkit/blob/bc92b63b98aa0968614240082997483f6bf68cbe/vendor/github.com/containerd/nydus-snapshotter/pkg/converter/convert_unix.go#L606
https://github.com/moby/buildkit/blob/bc92b63b98aa0968614240082997483f6bf68cbe/vendor/github.com/containerd/nydus-snapshotter/pkg/converter/convert_unix.go#L45
https://github.com/moby/buildkit/blob/bc92b63b98aa0968614240082997483f6bf68cbe/vendor/github.com/containerd/nydus-snapshotter/pkg/converter/convert_unix.go#L629-L633

And the new option allows for custom files to be specified through `AppendFiles` to be added; https://github.com/containerd/nydus-snapshotter/blob/v0.13.13/pkg/converter/convert_unix.go#L628-L644

```go
	bootstrapRa, err := local.OpenReader(targetBootstrapPath)
	if err != nil {
		return nil, errors.Wrap(err, "open bootstrap reader")
	}
	defer bootstrapRa.Close()

	files := append([]File{
		{
			Name:   EntryBootstrap,
			Reader: content.NewReader(bootstrapRa),
			Size:   bootstrapRa.Size(),
		},
	}, opt.AppendFiles...)
	var rc io.ReadCloser

	if opt.WithTar {
		rc = packToTar(files, false)
```

I guess my concern here ia; how can this behavior be controlled, as (IIUC, but I'm not super-familiar with all of this) this could result in arbitrary files to be injected?

Reply:
Hi @thaJeztah, [here](https://github.com/dragonflyoss/nydus/blob/master/contrib/nydusify/examples/manifest/manifest.json) is a nydus image manifest example, the manifest consists of a bootstrap layer (`containerd.io/snapshot/nydus-bootstrap`) and several blob layers (`containerd.io/snapshot/nydus-blob`).

- The bootstrap: it contains meta-information (inode, attrs...) about files and directories of all layers.
- The blob: is the data of each layer.

The option `AppendFiles` allows image build / conversion tools to append some files to the bootstrap layer, the format like:

```
gzip(tar(bootstrap + appended_files...))
```

so that nydus runtime (for example nydus snapshotter) can get more information about image, but these tools can't inject arbitrary files to image filesystem (bootstrap) by the converter package. :)